pkg/auth/handler/webapp/settings.go

```diff
@@ -68,91 +126,77 @@ func (h *SettingsHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	defer ctrl.Serve()
 
 	redirectURI := httputil.HostRelative(r.URL).String()
-	authenticatorID := r.Form.Get("x_authenticator_id")
+	providerAlias := r.Form.Get("x_provider_alias")
+	identityID := r.Form.Get("x_identity_id")
 	userID := ctrl.RequireUserID()
+	nonceSource, _ := r.Cookie(h.CSRFCookie.Name)
 
 	ctrl.Get(func() error {
-		data := map[string]interface{}{}
-
-		baseViewModel := h.BaseViewModel.ViewModel(r, nil)
-		viewmodels.Embed(data, baseViewModel)
-
-		authenticators, err := h.Authenticators.List(userID)
+		data, err := h.GetData(r, w)
 		if err != nil {
-			panic(err)
-		}
-
-		totp := false
-		oobotp := false
-		password := false
-		for _, typ := range h.Authentication.SecondaryAuthenticators {
-			switch typ {
-			case authn.AuthenticatorTypePassword:
-				password = true
-			case authn.AuthenticatorTypeTOTP:
-				totp = true
-			case authn.AuthenticatorTypeOOB:
-				oobotp = true
-			}
-		}
-
-		viewModel := SettingsViewModel{
-			Authenticators:           authenticators,
-			SecondaryTOTPAllowed:     totp,
-			SecondaryOOBOTPAllowed:   oobotp,
-			SecondaryPasswordAllowed: password,
+			return err
 		}
-		viewmodels.Embed(data, viewModel)
 
 		h.Renderer.RenderHTML(w, r, TemplateWebSettingsHTML, data)
 		return nil
 	})
 
-	ctrl.PostAction("revoke_devices", func() error {
-		if err := h.MFA.InvalidateAllDeviceTokens(userID); err != nil {
+	ctrl.PostAction("link_oauth", func() error {
+		opts := webapp.SessionOptions{
+			RedirectURI: redirectURI,
+		}
+		intent := intents.NewIntentAddIdentity(userID)
+
+		result, err := ctrl.EntryPointPost(opts, intent, func() (input interface{}, err error) {
+			input = &InputUseOAuth{
+				ProviderAlias:    providerAlias,
+				NonceSource:      nonceSource,
+				ErrorRedirectURI: httputil.HostRelative(r.URL).String(),
+			}
+			return
+		})
+		if err != nil {
 			return err
 		}
-		http.Redirect(w, r, redirectURI, http.StatusFound)
+
+		result.WriteResponse(w, r)
 		return nil
 	})
 
-	ctrl.PostAction("setup_secondary_password", func() error {
+	ctrl.PostAction("unlink_oauth", func() error {
```

Comment:
I found that when I unlink the last oauth identity, it shows me unexpected error in setting page. Not sure it is a old or new bug.